### PR TITLE
doc: update xline client document

### PR DESCRIPTION
**Pull request by bsbds**
This PR is based on #372

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
   
    descibed in `Update documents` section in #369

* what changes does this pull request make?

    Add more detailed descriptions to some methods and types.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline-client/src/clients/kv.rs [hunk 1]
```diff
@@ -86,22 +90,22 @@ impl KvClient {
         Ok(cmd_res.decode().into())
     }
 
-    /// Send `TxnRequest` by `CurpClient`
+    /// Creates a transaction, which can provide serializable reads and writes
```
Comment (Phoenix500526) [inline xline-client/src/clients/kv.rs:93]: serializable write? Txn doesn't provide serializable reads, range request does.

File: xline-client/src/clients/lease.rs [hunk 2]
```diff
@@ -77,22 +79,23 @@ impl LeaseClient {
         Ok(cmd_res.decode().into())
     }
 
-    /// Send `LeaseRevokeRequest` by `CurpClient`
+    /// Revokes a lease. All keys attached to the lease will expire and be deleted.
     ///
     /// # Errors
     ///
-    /// If client failed to send request
+    /// This function will return an error if the inner RPC client encountered a propose failure
```
Comment (Phoenix500526) [inline xline-client/src/clients/lease.rs:86]: It's "the inner CURP client", not "the inner RPC client". Please keep consistency with other comments.

File: xline-client/src/clients/lease.rs [hunk 3]
```diff
@@ -77,22 +79,23 @@ impl LeaseClient {
...
     }
 
-    /// Send `LeaseKeepAliveRequest` by `CurpClient`
+    /// Keeps the lease alive by streaming keep alive requests from the client
+    /// to the server and streaming keep alive responses from the server to the client.
     ///
     /// # Errors
     ///
-    /// If client failed to send request
+    /// This function will return an error if the inner RPC client encountered a propose failure
```
Comment (Phoenix500526) [inline xline-client/src/clients/lease.rs:98]: Same as above.

File: xline-client/src/clients/lease.rs [hunk 4]
```diff
@@ -122,11 +125,11 @@ impl LeaseClient {
         Ok((LeaseKeeper::new(id, sender), stream))
     }
 
-    /// Send `LeaseTimeToLiveRequest` by `CurpClient`
+    /// Retrieves lease information.
     ///
     /// # Errors
     ///
-    /// If client failed to send request
+    /// This function will return an error if the inner RPC client encountered a propose failure
```
Comment (Phoenix500526) [inline xline-client/src/clients/lease.rs:132]: Same as above

File: xline-client/src/clients/watch.rs [hunk 5]
```diff
@@ -40,11 +40,11 @@ impl WatchClient {
     ///
     /// # Errors
     ///
-    /// If the RPC client fails to send request
+    /// This function will return an error if the RPC client fails to send request
```
Comment (Phoenix500526) [inline xline-client/src/clients/watch.rs:43]: Same as above

File: xline-client/src/types/auth.rs [hunk 6]
```diff
@@ -9,7 +9,10 @@ pub struct AuthenticateRequest {
 }
 
 impl AuthenticateRequest {
-    /// New `AuthenticateRequest`
+    /// Creates a new `AuthenticateRequest`.
+    ///
+    /// `name` is the name of the user to authenticate,
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:14]: These two comments are a bit verbose. Just directly change the parameter names to `user_name` and `user_password`. Some places have parameter explanations, while others do not.

File: xline-client/src/types/auth.rs [hunk 7]
```diff
@@ -36,7 +39,9 @@ pub struct AuthUserAddRequest {
 }
 
 impl AuthUserAddRequest {
-    /// New `AuthUserAddRequest`
+    /// Creates a new `AuthUserAddRequest`.
+    ///
+    /// `name` is the name of the user to add.
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:44]: Same as above.

File: xline-client/src/types/auth.rs [hunk 8]
```diff
@@ -47,15 +52,15 @@ impl AuthUserAddRequest {
...
     pub fn with_pwd(mut self, password: impl Into<String>) -> Self {
         self.inner.password = password.into();
         self
     }
 
-    /// Set no password.
+    /// If set, Xline will add the user without password, password set with `with_pwd` will be ignored.
     #[inline]
     #[must_use]
     pub const fn with_no_pwd(mut self) -> Self {
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:66]: Is it mandatory for a user to set his password? If not, why don't we default to no password unless the user explicitly invokes `with_pwd`?

File: xline-client/src/types/auth.rs [hunk 9]
```diff
@@ -79,7 +84,9 @@ pub struct AuthUserGetRequest {
 }
 
 impl AuthUserGetRequest {
-    /// New `AuthUserGetRequest`
+    /// Creates a new `AuthUserGetRequest`.
+    ///
+    /// `name` is the name of the user to get
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:89]: Same as above

File: xline-client/src/types/auth.rs [hunk 10]
```diff
@@ -103,7 +110,9 @@ pub struct AuthUserDeleteRequest {
 }
 
 impl AuthUserDeleteRequest {
-    /// New `AuthUserDeleteRequest`
+    /// Creates a new `AuthUserDeleteRequest`.
+    ///
+    /// `name` is the name of the user to delete.
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:115]: Same as above

File: xline-client/src/types/auth.rs [hunk 11]
```diff
@@ -127,7 +136,10 @@ pub struct AuthUserChangePasswordRequest {
 }
 
 impl AuthUserChangePasswordRequest {
-    /// New `AuthUserChangePasswordRequest`
+    /// Creates a new `AuthUserChangePasswordRequest`.
+    ///
+    /// `name` is the name of the user to change password,
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:141]: Same as above, but using `new_password` instead.

File: xline-client/src/types/watch.rs [hunk 12]
```diff
@@ -131,7 +133,7 @@ impl WatchRequest {
         self
     }
 
-    /// Set `key` and `range_end` when with from key
+    /// If set, Xline will watch all keys that are equal or greater than the given key
```
Comment (Phoenix500526) [inline xline-client/src/types/watch.rs:136]: It's “equal to or greater than”.

File: xline-client/src/types/watch.rs [hunk 13]
```diff
@@ -142,55 +144,63 @@ impl WatchRequest {
...
+    /// `progress_notify` is set so that the Xline server will periodically send a `WatchResponse` with no events to the new watcher if there are no recent events. It is useful when clients wish to recover a disconnected watcher starting from a recent known revision. The xline server may decide how often it will send notifications based on current load.
     #[inline]
     #[must_use]
     pub const fn with_progress_notify(mut self) -> Self {
         self.inner.progress_notify = true;
         self
     }
 
-    /// Set `filters`
+    /// filters filter the events at server side before it sends back to the watcher.
```
Comment (Phoenix500526) [inline xline-client/src/types/watch.rs:175]: `filters` filter the events on the server side before it sends back to the watcher. The filter should be decorated with a backtick and the correct expression is "on XX side", not "at XX side"

File: xline-client/src/types/watch.rs [hunk 14]
```diff
@@ -142,55 +144,63 @@ impl WatchRequest {
...
+    /// If the previous KV is already compacted, nothing will be returned.
     #[inline]
     #[must_use]
     pub const fn with_prev_kv(mut self) -> Self {
         self.inner.prev_kv = true;
         self
     }
 
-    /// Set `watch_id`
+    /// If `watch_id` is provided and non-zero, it will be assigned to this watcher.
```
Comment (Phoenix500526) [inline xline-client/src/types/watch.rs:192]: Non-zero indicates that a user can use a negative number as the watch_id?

File: xline-client/src/types/kv.rs [hunk 15]
```diff
@@ -146,31 +155,34 @@ impl RangeRequest {
...
+    /// If `range_end` is '\0', the range is all keys >= key.
     #[inline]
     #[must_use]
     pub fn with_range_end(mut self, range_end: impl Into<Vec<u8>>) -> Self {
         self.inner.range_end = range_end.into();
         self
     }
 
-    /// Set `limit`
+    /// limit is a limit on the number of keys returned for the request. When limit is set to 0,
```
Comment (Phoenix500526) [inline xline-client/src/types/kv.rs:167]: The "limit" should be enclosed in a pair of backticks

File: xline-client/src/types/kv.rs [hunk 16]
```diff
@@ -146,31 +155,34 @@ impl RangeRequest {
...
+    /// it is treated as no limit.
     #[inline]
     #[must_use]
     pub fn with_limit(mut self, limit: i64) -> Self {
         self.inner.limit = limit;
         self
     }
 
-    /// Set `revision`
+    /// revision is the point-in-time of the key-value store to use for the range.
```
Comment (Phoenix500526) [inline xline-client/src/types/kv.rs:176]: Same as above

File: xline-client/src/types/kv.rs [hunk 17]
```diff
@@ -375,7 +398,7 @@ impl DeleteRangeRequest {
         self
     }
 
-    /// Set `key` and `range_end` when with from key
+    /// If set, Xline will delete all keys that are equal or greater than the given key
```
Comment (Phoenix500526) [inline xline-client/src/types/kv.rs:401]: equal to or greater than

File: xline-client/src/clients/lease.rs (hunk 2 above)
Comment (bsbds) [reply xline-client/src/clients/lease.rs:86]: the `revoke` method actually use the lease RPC client and does not use CURP client.

File: xline-client/src/types/watch.rs (hunk 14 above)
Comment (bsbds) [reply xline-client/src/types/watch.rs:192]: yes, there's no constraint that the watch_id must be non-negative in both etcd and xline.

**Review by themanforfree — COMMENTED** [review]
`api_type`, `lease_gen` and `AuthService` should not public to users

File: xline-client/src/types/auth.rs (hunk 6 above)
Comment (bsbds) [reply xline-client/src/types/auth.rs]: done

File: xline-client/src/types/auth.rs (hunk 8 above)
Comment (bsbds) [reply xline-client/src/types/auth.rs]: makes sense, I removed the with_no_pwd method and set it as default.

**Comment by bsbds** [comment]
> `api_type`, `lease_gen` and `AuthService` should not public to users

fixed, and I moved some of the re-exported api types to `src/types`

**Review by Phoenix500526 — DISMISSED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**